File: agent/transcript.py
from youtube_transcript_api import YouTubeTranscriptApi

from cache import transcript_cache


def get_transcript(video_id: str) -> str | None:
    cached = transcript_cache.get(video_id)
    if cached is not None:
        return cached

    try:
        entries = YouTubeTranscriptApi.get_transcript(video_id)
        text = " ".join(e["text"] for e in entries)
        transcript_cache.set(video_id, text)
        return text
    except Exception:
        return None


def get_transcripts(video_ids: list[str]) -> dict[str, str | None]:
    results = {}
    for vid in video_ids:
        results[vid] = get_transcript(vid)
    return results

File: api/transcript.js
const { YoutubeTranscript } = require("youtube-transcript");

async function getTranscript(videoId) {
  try {
    const entries = await YoutubeTranscript.fetchTranscript(videoId);
    if (!entries || entries.length === 0) return null;
    return entries.map((e) => e.text).join(" ");
  } catch {
    return null;
  }
}

async function getTranscripts(videoIds) {
  const results = {};
  const batches = [];
  for (let i = 0; i < videoIds.length; i += 5) {
    batches.push(videoIds.slice(i, i + 5));
  }
  for (const batch of batches) {
    await Promise.all(
      batch.map(async (id) => {
        results[id] = await getTranscript(id);
      })
    );
  }
  return results;
}

module.exports = { getTranscript, getTranscripts };
